[Janice Chen] iP

janjanchen opened the pull request:
# DukePro

> Made with a project template for a greenfield Java project. It's named after the Java mascot _Duke_. 

### What can DukePro do?

DukePro is your companion to free your mind of having to remember things you need to do! It's:
- Text-Based
- Easy to learn and use
- **_VERY_** efficient and handy!


### How to set up DukePro?
Follow these steps to get your own DukePro:
1. Download it from [here](https://github.com/janjanchen/ip). 
2. Run it. 
3. Add your tasks. 
4. ~~Manage~~ **Let _DukePro_ manage** your tasks for you! :wink:


**Features**

- [X] Manages Todo Lists
- [X] Manages Deadlines
- [X] Manages Events
- [x] Save, Delete, Filters and many more!
- [ ] More features to be added

## Code SneakPeaks
This is the code in DukePro that can be found in its `Task` class:
```java
/**
 * A Parent class for the different types of input to the Task List.
 */
public class Task {
   protected String description;
   protected boolean isDone;

   /**
    * A public constructor to create a task.
    * @param description The description of the task.
    */
   public Task(String description) {
      this.description = description;
      this.isDone = false;
   }

   /**
    * Returns the string representation of the task's status.
    * @return the string representation of the task's status.
    */
   public String getStatusIcon() {
      return (isDone ? "[X] " : "[ ] "); // mark done task with X
   }

   /**
    * Marks the task as completed.
    */
   public void markAsDone() {
      this.isDone = true;
   }

   /**
    * Returns the string representation of the task.
    * @return the string representation of the task.
    */
   @Override
   public String toString() {
      return this.getStatusIcon() + this.description;
   }

   /**
    * Returns the string representation of the task
    * to be saved into the file.
    * @return the string representation of the task
    * to be saved into the file.
    */
   public String toStore() {
      return this.getStatusIcon() + this.description;
   }
}
```

janjanchen committed:
Level-1

janjanchen committed:
Level-2

janjanchen committed:
Level-3

janjanchen committed:
Level-4

janjanchen committed:
Level-4

janjanchen committed:
Level-5

janjanchen committed:
Level-6

janjanchen committed:
Level-6.0 Added more exception handlings

janjanchen committed:
A-TextUiTesting

janjanchen committed:
Exception Handlers

janjanchen committed:
Level-7

janjanchen committed:
Level-8

janjanchen committed:
Intermediate

janjanchen committed:
Intermediate-1

janjanchen committed:
Merge branch 7 and 8

janjanchen committed:
A-MoreOOP

janjanchen committed:
A-Packages

janjanchen committed:
A-Junit

janjanchen committed:
A-Jar

janjanchen committed:
A-JavaDoc

janjanchen committed:
A-CodingStandard

janjanchen committed:
Level-9

janjanchen committed:
Merge branch 'branch-A-JavaDoc'

janjanchen committed:
Merge branch 'branch-A-CodingStandard'

janjanchen committed:
Level-9

janjanchen committed:
Level-9.1

janjanchen committed:
Level-9.2

janjanchen committed:
Update GFMD

janjanchen committed:
Update GFMD v1.1

janjanchen committed:
Update GFMD v1.2

rish-16 commented on src/main/java/Duke/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+
+/**
+ * A Task that can be added to the Task List
+ * which has a deadline date after the description.
```
Public classes must have `@author` and `@version` information in the JavaDoc comment. So something like this:

```java
/**
 *
 * A Task that can be added to the Task List
 * which has a deadline date after the description.
 *
 * @author Janice Chen
 * @version CS2103 AY21/22 Semester 1
 * 
*/
```

bklimey replied:
Sorry, I think `@author` and `@version` is optional for this module as seen in this [minimal JavaDoc guildline](https://nus-cs2103-ay2122s1.github.io/website/schedule/week3/topics.html#W3-3b)

rish-16 commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,234 @@
+     * @param input The keyword.
+     * @throws InvalidFormatException Thrown when the format of the query is invalid.
+     */
+    public void find(String input) throws InvalidFormatException {
+        if (input.length() > 5) {
+            String keyword = input.substring(5);
+            ArrayList<Task> full = tasks.inArrayList();
+            ArrayList<Task> filtered = new ArrayList<>();
+            int len = tasks.arrSize();
+            for (int i = 0; i < len; i++) {
+                if (full.get(i).toString().contains(keyword)) {
+                    filtered.add(full.get(i));
+                }
+            }
+
+            if (filtered.size() < 1) {
+                System.out.println("There are no matching results!");
+            } else {
+                int len2 = filtered.size();
+                System.out.println("Here are the matching tasks in your list:");
+                for (int i = 1; i < len2 + 1; i++) {
+                    System.out.println(i + ". " + filtered.get(i - 1));
+                }
+            }
+        } else {
+            throw new InvalidFormatException("Please specify the keyword to be searched!");
+        }
+    }
```
```suggestion
Can try adding some in-line comments to guide reviewers on what a specific block is doing within each method.
```

rish-16 commented on src/main/java/Duke/Storage.java:
```diff
@@ -0,0 +1,113 @@
+     * @throws IOException Thrown when the file is not found.
+     */
+    public ArrayList<Task> load() throws IOException {
+        ArrayList<Task> arr = new ArrayList<>();
```
Rename the `arr` variable to something more descriptive? Like `fileContents` for example

rish-16 reviewed (commented):
LGTM! A few minor changes here and there but code is functional to me :)

Code is overall easy to read and follow along. Follows the coding standard for the most part. Consider commenting your code more inside the blocks beyond simple JavaDocs.

janjanchen committed:
Add Gradle

janjanchen committed:
Merge branch 'add-gradle-support'

janjanchen committed:
branch-Level-10

janjanchen committed:
Merge branch 'branch-Level-10'

janjanchen committed:
Level 10.1

GabrielWLM commented:
Nice work! Your code looks well organised and easy to read.

Besides a few minor edits I agree with from the previous reviewer, I have yet to find anything wrong with your code. Good job!

bklimey commented on src/main/java/Duke/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+ * A Task that can be added to the Task List
+ * which has a deadline date after the description.
+ */
+
```
Minor nits, there is no space between JavaDoc and Class definition. It appears in Duke.java as well.

May refer to our module [guidelines](https://se-education.org/guides/conventions/java/intermediate.html#comments) on comments as well.
>No blank line between the documentation block and the method/class

bklimey commented on src/main/java/Duke/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+    protected LocalDate by;
+
+    /**
+     * A public constructor to create a Deadline Task.
```
Minor coding conventions: There is a blank comment line between the descriptions and block tags. This applies to other Javadoc too.
>- Insert a blank comment line between the description and the list of tags, as shown. 

From [How to Write Doc Comments for the Javadoc Tool -Oracle](https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html#format)

bklimey commented on src/main/java/Duke/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+    @Override
+    public String toString() {
+        return "[D]" + super.toString()
+                + "--(by: " + by.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
```
I like that you wrap your length line to improve readability! 👍

bklimey commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+        try {
+            tasks = new TaskList(storage.load());
+            parser = new Parser(tasks, storage);
+            recieve = new Recieve(parser);
```
Minor spelling error: `recieve` → `receive`
In IntelliJ, you could refactor>rename to mass rename the variable name.

bklimey commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    String welcomeMessage() {
+        return "Hello! I'm Duke\n" +
```
I like that you added instructions on how to use! 👍

bklimey commented on src/main/java/Duke/DukeException.java:
```diff
@@ -0,0 +1,26 @@
+
+
+    /**
+     * Returns the string representation of Duke Exception.
```
I feel that this may be optional as it adds limited value to readers.

From [guideline](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#do-not-repeat-the-obvious):
> If the code is self-explanatory, refrain from repeating the description in a comment just for the sake of 'good documentation'.

bklimey commented on src/main/java/Duke/InvalidFormatException.java:
```diff
@@ -0,0 +1,25 @@
+ * An exception thrown in Duke when
+ * the format of the user's input is invalid.
+ */
+public class InvalidFormatException extends Exception{
```
I like that you added your own Exception! 👍

bklimey commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    /**
+     * Filtes the tasks that includes the keyword.
```
Typo: `Filtes` → `Filters`

bklimey reviewed (commented):
Generally, I do not see any major issues. However, the readibility of the code could be improved. I saw some complicated expressions that can be broken down into [more readable steps](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-complicated-expressions). I too, had complicated expressions in my code and I found it hard to mantain and debug, and I have spent some time to clean them up. If you have time, maybe you could also work on the documentations. All the best! :)

janjanchen committed:
Add Assertions

Enabled assertion in Duke() methods

There were no assertions included in Duke() previously and assertions were disabled.

Added assertions in Duke() constructor, methods in TaskList class and handleUserInput() in MainWindow.

Assertions are added to verify assumptions made for the related methods before Runtime Exception are encounted.

janjanchen committed:
Restructure Code Logic

Code Logic/ Quality can be improved.

Restructure the code logic flow for better readability, which can enhance code quality.

Deeply nested if-else statements are removed. Use of more guard clauses to make 'happy path' more prominent.

janjanchen committed:
Merge pull request #1 from janjanchen/branch-A-Assertions

Merge to add A-Assertions

janjanchen committed:
Merge pull request #2 from janjanchen/branch-A-CodeQuality

Merge to add A-CodeQuality

janjanchen committed:
Add Date Time fields

Add date and time fields to Deadline and Event classes.

Event: There is start date/ time and end date/ time fields.
Deadline: There is time filed for the deadline.

Date and Time fields are added for a more comprehensive and clearer details on each task added to the list.

Change the store, write and read method after addition of new fields.

janjanchen committed:
Implement sort method

janjanchen committed:
Add First Welcome Message

janjanchen committed:
Add Sub-heading to GFMD

janjanchen committed:
Add Sub-heading to GFMD

janjanchen committed:
Design GUI

janjanchen committed:
Change dialog box color

janjanchen committed:
Add User Guide

janjanchen committed:
Add Ui.png

janjanchen committed:
Set theme jekyll-theme-cayman

janjanchen committed:
Edit Ui.png

janjanchen committed:
Merge remote-tracking branch 'origin/master'

janjanchen committed:
Edit Ui.png

janjanchen committed:
Rename package and split long methods to shorter methods

janjanchen committed:
Add JavaDocs and fix typos
